File: flutter_app/lib/services/location_service.dart
import 'dart:math';
import 'package:flutter/foundation.dart';
import 'package:geolocator/geolocator.dart';
import 'package:geocoding/geocoding.dart';
import '../data/helpline_data.dart';
import '../models/helpline.dart';

class LocationService extends ChangeNotifier {
  String _currentState = "Uttar Pradesh";
  String _district = "Lucknow";
  double _latitude = 26.8467;
  double _longitude = 80.9462;
  bool _isLiveGPS = false;
  bool _isLoading = false;

  String get currentState => _currentState;
  String get district => _district;
  double get latitude => _latitude;
  double get longitude => _longitude;
  bool get isLiveGPS => _isLiveGPS;
  bool get isLoading => _isLoading;

  StateInfo get currentStateInfo =>
      kIndiaStates[_currentState] ?? kIndiaStates["Uttar Pradesh"]!;

  String get mapsUrl =>
      "https://maps.google.com/?q=${_latitude.toStringAsFixed(6)},${_longitude.toStringAsFixed(6)}";

  void setStateManual(String stateName) {
    if (kIndiaStates.containsKey(stateName)) {
      final s = kIndiaStates[stateName]!;
      _currentState = stateName;
      _district = s.capital;
      _latitude = s.lat;
      _longitude = s.lng;
      _isLiveGPS = false;
      notifyListeners();
    }
  }

  Future<bool> requestLiveGPS() async {
    _isLoading = true;
    notifyListeners();

    try {
      bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
      if (!serviceEnabled) {
        _isLoading = false;
        notifyListeners();
        return false;
      }

      LocationPermission permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        permission = await Geolocator.requestPermission();
        if (permission == LocationPermission.denied) {
          _isLoading = false;
          notifyListeners();
          return false;
        }
      }

      if (permission == LocationPermission.deniedForever) {
        _isLoading = false;
        notifyListeners();
        return false;
      }

      Position position = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
        timeLimit: const Duration(seconds: 10),
      );

      _latitude = position.latitude;
      _longitude = position.longitude;
      _isLiveGPS = true;

      // Reverse geocoding
      try {
        List<Placemark> placemarks = await placemarkFromCoordinates(
          _latitude,
          _longitude,
        );
        if (placemarks.isNotEmpty) {
          final place = placemarks.first;
          final stateRaw = place.administrativeArea ?? "";
          final matched = _matchState(stateRaw);
          if (matched != null) {
            _currentState = matched;
          } else {
            _currentState = _findNearestState(_latitude, _longitude);
          }
          _district = place.subAdministrativeArea ?? place.locality ?? place.subLocality ?? kIndiaStates[_currentState]?.capital ?? "Area";
        }
      } catch (e) {
        // Fallback nearest state calculation
        _currentState = _findNearestState(_latitude, _longitude);
        _district = kIndiaStates[_currentState]?.capital ?? "Area";
      }

      _isLoading = false;
      notifyListeners();
      return true;
    } catch (e) {
      _isLoading = false;
      notifyListeners();
      return false;
    }
  }

  String? _matchState(String raw) {
    final clean = raw.toLowerCase().replaceAll(RegExp(r'state of|union territory of|nct of'), '').trim();
    for (final s in kIndiaStates.keys) {
      if (s.toLowerCase() == clean || clean.contains(s.toLowerCase())) {
        return s;
      }
    }
    return null;
  }

  String _findNearestState(double lat, double lng) {
    String closest = "Uttar Pradesh";
    double minDistance = double.infinity;

    kIndiaStates.forEach((stateName, info) {
      final dLat = lat - info.lat;
      final dLng = lng - info.lng;
      final dist = sqrt(dLat * dLat + dLng * dLng);
      if (dist < minDistance) {
        minDistance = dist;
        closest = stateName;
      }
    });

    return closest;
  }
}

File: js/location_service.js
// Location Service with Native GPS, Reverse Geocoding & Offline State Boundary Matcher
import { INDIA_STATES_DATA } from './data.js';

export class LocationService {
  constructor() {
    this.currentLocation = {
      latitude: 26.8467,
      longitude: 80.9462,
      state: "Uttar Pradesh",
      district: "Lucknow",
      formattedAddress: "Lucknow, Uttar Pradesh, India",
      accuracy: null,
      isLiveGPS: false,
      timestamp: null
    };

    // Load saved state preference if present
    const savedState = localStorage.getItem("helplines_active_state");
    if (savedState && INDIA_STATES_DATA[savedState]) {
      this.currentLocation.state = savedState;
      this.currentLocation.district = INDIA_STATES_DATA[savedState].capital;
      this.currentLocation.latitude = INDIA_STATES_DATA[savedState].lat;
      this.currentLocation.longitude = INDIA_STATES_DATA[savedState].lng;
    }

    this.listeners = [];
  }

  onLocationChange(callback) {
    this.listeners.push(callback);
  }

  notifyListeners() {
    for (const listener of this.listeners) {
      try {
        listener(this.currentLocation);
      } catch (e) {
        console.error("Error in location listener:", e);
      }
    }
  }

  /**
   * Requests real GPS coordinates from browser Geolocation API
   */
  async requestGPS() {
    return new Promise((resolve, reject) => {
      if (!("geolocation" in navigator)) {
        reject(new Error("Geolocation is not supported by your browser."));
        return;
      }

      navigator.geolocation.getCurrentPosition(
        async (position) => {
          const lat = position.coords.latitude;
          const lng = position.coords.longitude;
          const accuracy = Math.round(position.coords.accuracy);

          this.currentLocation.latitude = lat;
          this.currentLocation.longitude = lng;
          this.currentLocation.accuracy = accuracy;
          this.currentLocation.isLiveGPS = true;
          this.currentLocation.timestamp = new Date();

          // Try reverse geocoding to obtain exact State, District, Address
          const geoInfo = await this.resolveCoordinates(lat, lng);
          this.currentLocation.state = geoInfo.state;
          this.currentLocation.district = geoInfo.district;
          this.currentLocation.formattedAddress = geoInfo.formattedAddress;

          localStorage.setItem("helplines_active_state", this.currentLocation.state);
          this.notifyListeners();
          resolve(this.currentLocation);
        },
        (error) => {
          let msg = "Could not access location.";
          switch (error.code) {
            case error.PERMISSION_DENIED:
              msg = "Location permission was denied. You can select your state manually from the top bar.";
              break;
            case error.POSITION_UNAVAILABLE:
              msg = "Location information is currently unavailable.";
              break;
            case error.TIMEOUT:
              msg = "The request to get user location timed out.";
              break;
          }
          reject(new Error(msg));
        },
        {
          enableHighAccuracy: true,
          timeout: 10000,
          maximumAge: 60000
        }
      );
    });
  }

  /**
   * Resolves coordinates to State and District
   * Uses online API if reachable, falls back to offline nearest-centroid & bounding box
   */
  async resolveCoordinates(lat, lng) {
    // 1. Try online Nominatim Reverse Geocoding
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 3500);

      const res = await fetch(
        `https://nominatim.openstreetmap.org/reverse?format=json&lat=${lat}&lon=${lng}&zoom=10&addressdetails=1`,
        {
          signal: controller.signal,
          headers: { 'Accept': 'application/json' }
        }
      );
      clearTimeout(timeoutId);

      if (res.ok) {
        const data = await res.json();
        const address = data.address || {};
        const stateRaw = address.state || address.province || address.state_district;
        const districtRaw = address.county || address.state_district || address.city || address.town || address.village || "";

        const matchedState = this.matchStateName(stateRaw);
        if (matchedState) {
          return {
            state: matchedState,
            district: districtRaw,
            formattedAddress: data.display_name || `${districtRaw}, ${matchedState}`
          };
        }
      }
    } catch (e) {
      console.warn("Online reverse geocoding skipped or offline, using fallback geometry.", e);
    }

    // 2. Offline Fallback: Check bounding boxes and nearest centroids of all Indian states
    const fallbackState = this.matchCoordinatesOffline(lat, lng);
    const stateInfo = INDIA_STATES_DATA[fallbackState] || INDIA_STATES_DATA["Uttar Pradesh"];

    return {
      state: fallbackState,
      district: stateInfo.capital,
      formattedAddress: `Near ${stateInfo.capital}, ${fallbackState} (GPS: ${lat.toFixed(4)}, ${lng.toFixed(4)})`
    };
  }

  /**
   * Finds matching state by raw string (handles "NCT of Delhi", "State of Uttar Pradesh", etc.)
   */
  matchStateName(rawName) {
    if (!rawName) return null;
    const clean = rawName.toLowerCase().replace(/state of|nct of|union territory of/g, '').trim();

    for (const stateName of Object.keys(INDIA_STATES_DATA)) {
      if (stateName.toLowerCase() === clean || clean.includes(stateName.toLowerCase())) {
        return stateName;
      }
    }
    return null;
  }

  /**
   * Offline Geometric Matcher: calculates Euclidean distance to State centroids and checks bounds
   */
  matchCoordinatesOffline(lat, lng) {
    let closestState = "Uttar Pradesh";
    let minDistance = Infinity;

    for (const [stateName, data] of Object.entries(INDIA_STATES_DATA)) {
      // Check bounding box if provided
      if (data.bounds) {
        if (
          lat >= data.bounds.minLat &&
          lat <= data.bounds.maxLat &&
          lng >= data.bounds.minLng &&
          lng <= data.bounds.maxLng
        ) {
          return stateName;
        }
      }

      // Check distance to centroid
      const dLat = lat - data.lat;
      const dLng = lng - data.lng;
      const dist = Math.sqrt(dLat * dLat + dLng * dLng);
      if (dist < minDistance) {
        minDistance = dist;
        closestState = stateName;
      }
    }

    return closestState;
  }

  /**
   * Manually sets the active state (e.g. from state picker modal)
   */
  setState(stateName) {
    if (INDIA_STATES_DATA[stateName]) {
      const data = INDIA_STATES_DATA[stateName];
      this.currentLocation.state = stateName;
      this.currentLocation.district = data.capital;
      this.currentLocation.latitude = data.lat;
      this.currentLocation.longitude = data.lng;
      this.currentLocation.formattedAddress = `${data.capital}, ${stateName}`;
      this.currentLocation.isLiveGPS = false;

      localStorage.setItem("helplines_active_state", stateName);
      this.notifyListeners();
    }
  }

  /**
   * Generates emergency Google Maps link for the current coordinates
   */
  getMapsLink() {
    return `https://maps.google.com/?q=${this.currentLocation.latitude.toFixed(6)},${this.currentLocation.longitude.toFixed(6)}`;
  }
}
